### PR TITLE
dev: use upstream watch & notify actor

### PR DESCRIPTION
**Pull request by Myriad-Dreamin**
This problem is found in various editors, such as vscode and neovim.

### CODE REVIEW

File: src/actor/typst.rs [hunk 1]
```diff
@@ -47,6 +53,42 @@ type MpScChannel<T> = (mpsc::UnboundedSender<T>, mpsc::UnboundedReceiver<T>);
 type WatchChannel<T> = (watch::Sender<T>, watch::Receiver<T>);
 type BroadcastChannel<T> = (broadcast::Sender<T>, broadcast::Receiver<T>);
 
+#[derive(Debug, Clone)]
+struct SuperpositionDocument {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:57]: What does `Superposition` mean?

File: src/actor/typst.rs [hunk 2]
```diff
@@ -98,44 +140,98 @@ impl TypstActor {
...
             panic!();
         };
+
+        let mut pre_doc_receiver = self.pre_doc_receiver.take().unwrap();
+        let doc_sender = self.doc_sender.take().unwrap();
+        let renderer_sender = self.renderer_sender.take().unwrap();
+        std::thread::Builder::new()
+            .name("pre_receiver".to_owned())
+            .spawn(|| {
+                tokio::runtime::Builder::new_current_thread()
```
Comment (Enter-tainer) [inline src/actor/typst.rs:173]: ~~pls don't build a new runtime or you have very decent reason to do this~~ This seems to be OK

**Comment by Myriad-Dreamin** [comment]
https://github.com/Enter-tainer/typst-preview/suites/16544994568/artifacts/945124668

File: src/actor/typst.rs [hunk 3]
```diff
@@ -70,88 +59,116 @@ impl TypstActor {
...
-        else {
-            error!("TypstActor: failed to watch filesystem events");
-            panic!();
-        };
-        loop {
-            let mut recompile = false;
+
+    pub async fn run(self) {
+        let (server, client) = self.inner.split();
+        tokio::spawn(server.spawn());
```
Comment (Enter-tainer) [inline src/actor/typst.rs:95]: IIRC there no need to wrap `server.spawn()` with tokio::spawn.

File: src/actor/typst.rs [hunk 4]
```diff
@@ -70,88 +59,116 @@ impl TypstActor {
...
+        tokio::spawn(async move {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:96]: same here

File: src/actor/typst.rs [hunk 5]
```diff
@@ -70,88 +59,116 @@ impl TypstActor {
...
+                client.process_mail(mail).await;
             }
+            info!("TypstActor: exiting");
+        })
+        .await
+        .unwrap();
+    }
+}
+
+struct TypstClient<T> {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:110]: what is T & `replace` for? I think once cell should be suitable for this case

**Comment by Enter-tainer** [comment]
Let's merge this and wait for v0.9.0

**Pull request merged**